sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs

```diff
@@ -1,22 +1,40 @@
 // Copyright 2016-2020, Pulumi Corporation
 
-using System.Collections.Generic;
+using System;
+using System.Net;
+using System.Net.Http;
 using System.Threading.Tasks;
-using Grpc.Core;
+using Grpc.Net.Client;
 using Pulumirpc;
 
 namespace Pulumi
 {
     internal class GrpcEngine : IEngine
     {
         private readonly Engine.EngineClient _engine;
+        private static GrpcChannel? _engineChannel = null;
+        private readonly object _engineChannelLock = new object();
```

Comment:
Why is `_engineChannel` static? Should `_engineChannelLock` be static too? (Same questions for `GrpcMonitor` below).

Reply:
@justinvp According to the docs, creating a `GrpcChannel` is expensive whereas creating the gRPC client that uses the channel is not: hence I was keeping a single instance of the channel to be used across client instances to avoid recreating the channels every time. 

As for `_engineChannelLock` it doesn't need to be static since it is only used for locking. 

The lock however is no longer needed! I refactored the code to use a (concurrent) dictionary to keep track of the channels created by different Engine and Monitor addresses. This dictionary is thread-safe and doesn't need special locking